src/libstd/sys/unix/mutex.rs

```diff
@@ -28,14 +28,18 @@ impl Mutex {
         //
         // A pthread mutex initialized with PTHREAD_MUTEX_INITIALIZER will have
         // a type of PTHREAD_MUTEX_DEFAULT, which has undefined behavior if you
-        // try to re-lock it from the same thread when you already hold a lock.
+        // try to re-lock it from the same thread when you already hold a lock
+        // (https://pubs.opengroup.org/onlinepubs/9699919799/functions/pthread_mutex_init.html).
         //
         // In practice, glibc takes advantage of this undefined behavior to
         // implement hardware lock elision, which uses hardware transactional
-        // memory to avoid acquiring the lock. While a transaction is in
+        // memory to avoid acquiring the lock.
+        // This is the case even if PTHREAD_MUTEX_DEFAULT == PTHREAD_MUTEX_NORMAL
```

Comment:
This might sound confusing since just a few lines below we set the mutex type to `PTHREAD_MUTEX_NORMAL`. What is actually happening is that `pthread_mutexattr_settype(PTHREAD_MUTEX_NORMAL)` disables lock elision (to conform to the standard), while `PTHREAD_MUTEX_INITIALIZER` doesn't.

Reply:
See [this comment](https://sourceware.org/git/?p=glibc.git;a=blob;f=nptl/pthread_mutexattr_settype.c;h=c11a391e46fa43ac7cd4f8ba1309a3354f15ba57;hb=HEAD) in glibc.

Reply:
I edited the comment, better like this?